neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/core/internal/ContentsList.java

```diff
@@ -0,0 +1,285 @@
+    private final PersistentEObject owner;
+
+    /**
+     * The containment features that are handled by this list.
+     */
+    @Nonnull
+    private final List<EStructuralFeature> features;
+
+    /**
+     * {@code true} if proxied {@link EObject objects} has to be resolved.
+     */
+    private final boolean resolve;
+
+    /**
+     * The cached non-resolving list.
+     *
+     * @see #basicList()
+     */
+    @Nonnull
+    private final LazyReference<ContentsList<E>> nonResolvingList;
+
+    /**
+     * Constructs a new {@code ContentsList}.
+     *
+     * @param owner    the owner of this list
+     * @param features the containment features that are handled by this list
+     * @param resolve  {@code true} if proxied {@link EObject objects} has to be resolved
+     */
+    protected ContentsList(PersistentEObject owner, List<EStructuralFeature> features, boolean resolve) {
+        this.owner = owner;
+        this.features = features;
+        this.resolve = resolve;
+        this.nonResolvingList = LazyReference.soft(() -> resolve ? new ContentsList<>(owner, features, false) : this);
+    }
+
+    /**
+     * Returns an empty {@code ContentsList}.
+     *
+     * @param <E> the type of elements in this list
+     *
+     * @return an empty list
+     */
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    public static <E extends EObject> ContentsList<E> empty() {
+        return (ContentsList<E>) EMPTY;
+    }
+
+    /**
+     * Creates a new {@code ContentsList} with the given {@code owner}.
+     *
+     * @param owner the owner of this list
+     * @param <E>   the type of elements in this list
+     *
+     * @return a new list
+     */
+    @Nonnull
+    public static <E extends EObject> ContentsList<E> newList(PersistentEObject owner) {
+        List<EStructuralFeature> features = owner.eClass().getEAllStructuralFeatures();
+        EStructuralFeature[] containments = EClassImpl.FeatureSubsetSupplier.class.cast(features).containments();
+
+        return nonNull(containments)
+                ? new ContentsList<>(owner, Arrays.asList(containments), true)
+                : ContentsList.empty();
+    }
+
+    /**
+     * Returns {@code true} if the feature of the {@code entry} must be included in the result, {@code false} if it
+     * should be ignored.
+     *
+     * @param entry the entry to test
+     *
+     * @return {@code true} if the feature of the {@code entry} must be included in the result, {@code false} if it
+     * should be ignored
+     */
+    protected boolean isIncludedEntry(FeatureMap.Entry entry) {
+        final EStructuralFeature feature = entry.getEStructuralFeature();
+
+        return nonNull(entry.getValue())
+                && EObjects.isReference(feature)
+                && EObjects.asReference(feature).isContainment();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    // FIXME Feature maps are not filtered: attributes are also included with this implementation
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Take the required action to fix the issue indicated by this comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1134)